parser/src/main/java/me/vertretungsplan/parser/IphisParser.java

```diff
@@ -0,0 +1,358 @@
+     */
+    private String api;
+
+    /**
+     * Shortcode for school
+     */
+    private String kuerzel;
+    /** */
+    private String jwt_key;
+
+    /**
+     * URL of the school website
+     */
+    private String website;
+
+    /**  */
+    private JSONArray grades;
+    /**  */
+    private JSONArray teachers;
+    /**  */
+    private String authToken;
+
+    public IphisParser(SubstitutionScheduleData scheduleData, CookieProvider cookieProvider) {
+        super(scheduleData, cookieProvider);
+        JSONObject data = scheduleData.getData();
+        try {
+            api = "https://" + data.getString(PARAM_URL) + "/remote/vertretungsplan/ssp";
+            kuerzel = data.getString(PARAM_KUERZEL);
+            website = data.getString(PARAM_WEBSITE);
+            jwt_key = data.getString(PARAM_JWT_KEY);
+        } catch (JSONException e) {
+            e.printStackTrace();
+        }
+    }
+
+    public SubstitutionSchedule getSubstitutionSchedule()
+            throws IOException, JSONException, CredentialInvalidException {
+        final SubstitutionSchedule substitutionSchedule = SubstitutionSchedule.fromData(scheduleData);
+
+        grades = getGrades();
+        teachers = getTeachers();
+        final JSONArray changes = getChanges();
+
+        substitutionSchedule.setClasses(getAllClasses());
+        substitutionSchedule.setTeachers(getAllTeachers());
+        substitutionSchedule.setWebsite(website);
+
+        parseIphis(substitutionSchedule, changes, grades, teachers);
+        return substitutionSchedule;
+    }
+
+    /**
+     * Returns authentication key
+     */
+    private String getAuthenticationHeader() throws JSONException, IOException, CredentialInvalidException {
+        final UserPasswordCredential userPasswordCredential = (UserPasswordCredential) credential;
+        final String username = userPasswordCredential.getUsername();
+        final String password = userPasswordCredential.getPassword();
+
+        if (authToken == null) {
+            JSONObject payload = new JSONObject();
+            payload.put("school", kuerzel);
+            payload.put("user", username);
+            payload.put("password", password);
+
+            httpPost(api + "/login", "UTF-8", payload.toString(), ContentType.APPLICATION_JSON);
+            final String token = httpPost(api + "/login", "UTF-8", payload.toString(), ContentType.APPLICATION_JSON);
+
+            try {
+                final String key = Base64.getEncoder().encodeToString(jwt_key.getBytes());
+                final Claims jwtToken = Jwts.parser().setSigningKey(key)
+                        .parseClaimsJws(token).getBody();
+                assert jwtToken.getSubject().equals("vertretungsplan.me");
+
+                authToken = token;
+                website = jwtToken.getIssuer();
+                if (username.equals("lehrer")) {
+                    getSubstitutionSchedule().setType(SubstitutionSchedule.Type.TEACHER);
```

Comment:
Wenn ich es richtig verstehe, rufen `getGrades()`, `getTeachers()` und `getChanges()` über `getJSONArray()` `getAuthenticationHeader()` auf. Wenn dann hier wieder `getSubstitutionSchedule()` aufgerufen wird, führt das doch zu endloser Rekursion, oder nicht?

Hier nochmal den Typ des Plans zu setzen ist auch nicht nötig, denn der Typ wird ja schon durch die übergebene `SubstitutionScheduleData` festgelegt. Es müsste aber vermutlich sichergestellt werden, dass nur der Lehrerbenutzer den Lehrerplan auswählen kann.

Reply:
Woher kommt dann die Info, ob es sich um einen Lehrerplan oder einen Schülerplan (oder werden in der App einfach ggf. zwei verschiedene Zugangsdaten abgefragt? - dann kann das natürlich weg!)

Reply:
In der App kann man auswählen ob man den Lehrer- oder den Schülerplan sehen will, bevor man die Zugangsdaten eingeben muss. Das sind dann zwei unterschiedliche Pläne mit eigenen `SubstitutionScheduleData`s.